gemfire-core/src/main/java/com/gemstone/gemfire/cache/GemFireSparkConnectorCacheFactory.java

```diff
@@ -1,46 +1,157 @@
+    synchronized (CacheFactory.class) {
+      DistributedSystem ds = null;
+      if (this.dsProps.isEmpty()) {
+        // any ds will do
+        ds = InternalDistributedSystem.getConnectedInstance();
+      }
+      if (ds == null) {
+        ds = DistributedSystem.connect(this.dsProps);
+      }
+      PoolFactory defaultPF = null;
+       String defaultGrid =  gfeGridMappings.remove(GemFireSparkConnectorCacheImpl.gfeGridPropPrefix);
+      if (defaultGrid != null) {
+        defaultPF = this.createAndConfigurePoolFactory(defaultGrid);
+      }
+      Cache cache = GemFireSparkConnectorCacheImpl.create(defaultPF, gfeGridMappings, ds, cacheConfig);
+
+      for (Map.Entry<String, String> gridEntry : gfeGridMappings.entrySet()) {
+        String key = gridEntry.getKey().trim();
+        String gridName = key.substring(key.indexOf('.') + 1);
+        String locators = gridEntry.getValue().trim();
+        PoolFactory pf = this.createAndConfigurePoolFactory(locators);
+        pf.create(gridName);
+      }
+      return cache;
+    }
+  }
+
+
+  private PoolFactory createAndConfigurePoolFactory(String remoteLocators) {
+    PoolFactory pf = PoolManager.createFactory();
+    pf.setReadTimeout(30000);
+
+    StringTokenizer remoteLocatorsTokenizer = new StringTokenizer(remoteLocators, ",");
+    DistributionLocatorId[] locators = new DistributionLocatorId[remoteLocatorsTokenizer
+        .countTokens()];
+    int i = 0;
+    while (remoteLocatorsTokenizer.hasMoreTokens()) {
+      locators[i++] = new DistributionLocatorId(remoteLocatorsTokenizer.nextToken());
+    }
+    List<ServerLocation> servers = new ArrayList<ServerLocation>();
+    for (DistributionLocatorId locator : locators) {
+      try {
+
+        InetSocketAddress addr = new InetSocketAddress(locator.getHost(), locator.getPort());
+        GetAllServersRequest req = new GetAllServersRequest("");
+        Object res = TcpClient.requestToServer(addr.getAddress(), addr.getPort(), req, 2000);
+        if (res != null) {
+          servers.addAll((List<ServerLocation>)((GetAllServersResponse)res).getServers());
+        }
+      } catch (Exception e) {
+        System.out.println("Unable to get remote gfe servers from locator = " + locator);
+      }
+    }
+
+    List<ServerLocation> prefServers = null;
+    if (servers.size() > 0) {
+      String sparkIp = System.getenv("SPARK_LOCAL_IP");
+      String hostName = null;
+      try {
+        hostName = sparkIp != null ? InetAddress.getByName(sparkIp).getCanonicalHostName() :
+            InetAddress.getLocalHost().getCanonicalHostName();
+      } catch (Exception e) {
+        hostName = "";
+      }
+      int spacing = servers.size() / 3;
+      if (spacing != 0) {
+        prefServers = new ArrayList<ServerLocation>();
+        ServerLocation localServer = null;
+        for (int j = 0; j < servers.size(); ++j) {
+          ServerLocation sl = servers.get(j);
+          if (localServer == null && sl.getHostName().equals(hostName)) {
```

Comment:
There are some convoluted ways like external websites that will determine the public IP/hostname from the HTTP translation headers. For snappydata clients right now we are not using that rather explicit "hostname-for-clients" boot property which has to be set to the public hostname.

Similar is there on GemFire side where the string passed to DistributionLocatorId has to have "hostname-for-clients" sub-property in the string -- see the DistributionLocatorId(String) constructor. I suppose you could assume that this sub-property has to be present in remote-locators property and matching on both sides. So you can pick up the DistributionLocatorId.getHostnameForClients() if it is non-null or else the SocketCreator.getLocalHost() if it is null.